File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> window = new HashSet<>();
        int n = s.length();
        
        int left = 0;
        int result = 0;
        
        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);
            while (window.contains(c)) {
                window.remove(s.charAt(left++));
            }
            
            window.add(c);
            
            result = Math.max(result, i - left + 1);
        }
        
        return result;
    }
}

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> charIndex;

        int answer = 0;
        int left = -1;
        for(int right = 0; right < s.length(); right++) {
            if(charIndex.find(s[right]) != charIndex.end() && charIndex[s[right]] > left) {
               left = charIndex[s[right]];
            }
            answer = max(answer, right - left);
            charIndex[s[right]] = right;
        }

        return answer;
    }
};